Add readr 2.0.0 blog post (#522)
tidyverse/tidyverse.org: content/blog/readr-2-0-0/index.Rmd

---
output: hugodown::hugo_document

slug: readr-2-0-0
title: readr 2.0.0
date: 2021-07-06
author: Jim Hester
description: |
  This major release of readr includes a new multi-threaded parsing engine powered by vroom and a 
  number of user interface improvements.
photo:
  url: https://unsplash.com/photos/XOW1WqrWNKg
  author: Anastasia Zhenina
categories:
  - package
---

We're thrilled to announce the release of [readr](https://readr.tidyverse.org/) 2.0.0!

The readr package makes it easy to get rectangular data out of comma separated
(csv), tab separated (tsv) or fixed width files (fwf) and into R. It
is designed to flexibly parse many types of data found in the wild, while still
cleanly failing when data unexpectedly changes.

The easiest way to install the latest version from CRAN is to install the whole tidyverse.

```{r eval = FALSE}
install.packages("tidyverse")
```

Alternatively, install just readr from CRAN:

```{r eval = FALSE}
install.packages("readr")
```

> This blog post will show off the most important changes to the package.
> 
> You can see a full list of changes in the [readr release notes](https://github.com/r-lib/readr/releases) and [vroom release notes](https://github.com/r-lib/vroom/releases).

Reviewer: ```suggestion
This blog post will show off the most important changes to the package. You can see a full list of changes in the [readr release notes](https://github.com/r-lib/readr/releases) and [vroom release notes](https://github.com/r-lib/vroom/releases).
```

Reviewer: Also needs rough description of the major changes

```{r setup}
library(readr)
```

> ## readr edition two

Reviewer: ```suggestion
## readr 2nd edition
```
?

> Readr 2.0.0 is a major release of readr and introduces a new second edition parsing and writing engine implemented via the [vroom](https://vroom.r-lib.org/) package.
> 
> This engine takes advantage of lazy reading, multi-threading and performance characteristics of modern SSD drives to significantly improve the performance of reading and writing compared to the first edition engine.
> 
> We will continue to support the first edition for a number of releases, but eventually this support will be first deprecated and then removed.
> 
> You can use the `with_edition()` or `local_edition()` functions to temporarily change the edition of readr for a section of code.
> 
> e.g.
> 
> - `with_edition(1, read_csv("my_file.csv"))` will read `my_file.csv` with the first edition of readr.
> 
> - `readr::local_edition(1)` placed at the top of your function or script will use the first edition for the rest of the function or script.

Reviewer: ```suggestion
readr 2.0.0 is a major release of readr and introduces a new second edition parsing and writing engine implemented via the [vroom](https://vroom.r-lib.org/) package. This engine takes advantage of lazy reading, multi-threading and performance characteristics of modern SSD drives to significantly improve the performance of reading and writing compared to the first edition engine. 

We have done our best to ensure that the two editions parse csv files as similarly as possible, but in case there are differences that affect your code, you can use the `with_edition()` or `local_edition()` functions to temporarily change the edition of readr for a section of code:

- `with_edition(1, read_csv("my_file.csv"))` will read `my_file.csv` with the first edition of readr.

- `readr::local_edition(1)` placed at the top of your function or script will use the first edition for the rest of the function or script.

We will continue to support the 1st edition for a number of releases, but our goal is to ensure that the 2nd edition is uniformly better than the 1st edition so we plan to eventually deprecate and then remove the 1st edition code. 
```

## Reading multiple files at once

> Edition two has built-in support for reading sets of files with the

Reviewer: I think 2nd edition is easier to read, so I'd suggest changing everywhere.

same columns into one output table in a single command.
Just pass the filenames to be read in the same vector to the reading function.

First we generate some files to read by splitting the nycflights dataset by
airline.

```{r}
library(nycflights13)
purrr::iwalk(
  split(flights, flights$carrier),
  ~ { .x$carrier[[1]]; vroom::vroom_write(.x, glue::glue("flights_{.y}.tsv"), delim = "\t") }
)
```

Then we can efficiently read them into one tibble by passing the filenames
directly to readr.

```{r}
files <- fs::dir_ls(glob = "flights*tsv")
files
readr::read_tsv(files)
```

If the filenames contain data, such as the date when the sample was collected,

> use `id` argument to include the paths as a column in the data.

Reviewer: Show example?

You will likely have to post-process the paths to keep only the relevant portion for your use case.

## Delimiter guessing

> Edition two supports automatic guessing of delimiters.

Reviewer: Credit data.table for inspiration

Because of this you can now use `read_delim()` without specifying a `delim` argument in many cases.

```{r}
x <- read_delim(readr_example("mtcars.csv"))
```

## New column specification output

On Feburary 11, 2021 we conducted a [survey on twitter](https://twitter.com/jimhester_/status/1359969288501739528) asking for the community's opinion on the column specification output.
We recieved over 750 responses to the survey!
It revealed some useful information.

- 3/4 of respondents found printing the column specifications helpful.

> - 2/3 of respondents preferred the edition output vs legacy output.

Reviewer: ```suggestion
- 2/3 of respondents preferred the vroom output over the readr 1.0.0 output.
```
?

- Only 1/5 of respondents correctly knew how to supress printing of the column specifications.

Based on these results we have added two new ways to more easily suppress the column specification printing.

- Use `read_csv(show_col_types = FALSE)` to disable printing for a single function call.
- Use `options(readr.show_types = FALSE)` to disable printing for the entire session.

We will also continue to print the column specifications with the new output.

Note you can still obtain the old output style by printing the column specification object directly.

```{r}
spec(x)
```

> Or the new style by calling `summary()` on the specficiation object.

Reviewer 2: ```suggestion
Or the new style by calling `summary()` on the specification object.
```

```{r}
summary(spec(x))
```

## Column selection

> Edition two introduces a new argument, `col_select`, which makes selecting columns to
> keep (or omit) more straightforward than before.
> 
>  `col_select` uses the same interface as `dplyr::select()`, so you can perform very flexible selection operations.
> 
> Select with the column names directly.

Reviewer: ```suggestion
Edition two introduces a new argument, `col_select`, which makes selecting columns to
keep (or omit) more straightforward than before.  `col_select` uses the same interface as `dplyr::select()`, so you can perform very flexible selection operations:

* Select with the column names directly.
```

And make the whole thing a bulleted list with one bullet for each case

```{r}
data <- read_tsv("flights_AA.tsv", col_select = c(year, flight, tailnum))
```

Or by numeric column.
```{r}
data <- read_tsv("flights_AA.tsv", col_select = c(1, 2))
```

Drop columns by name by prefixing them with `-`.
```{r}

> data <- read_tsv("flights_AA.tsv", col_select = c(-dep_time, -air_time:-time_hour))

Reviewer: ```suggestion
data <- read_tsv("flights_AA.tsv", col_select = c(-dep_time, -(air_time:time_hour)))
```
?

```

Use the selection helpers such as `ends_with()`.
```{r}
data <- read_tsv("flights_AA.tsv", col_select = ends_with("time"))
```

Or even rename columns by using a named list.
```{r}
data <- read_tsv("flights_AA.tsv", col_select = list(plane = tailnum, everything()))
data
```

## Name repair

Often the names of columns in the original dataset are not ideal to work with.
Edition two uses the same [name_repair](https://www.tidyverse.org/articles/2019/01/tibble-2.0.1/#name-repair)
argument as in the tibble package, so you can use one of the default name repair strategies or
provide a custom function.
One useful approach is to use the
[janitor::make_clean_names()](http://sfirke.github.io/janitor/).

```{r, message = FALSE}
read_tsv("flights_AA.tsv", name_repair = janitor::make_clean_names)

read_tsv("flights_AA.tsv", name_repair = ~ janitor::make_clean_names(., case = "lower_camel"))
```

```{r, show = FALSE}
unlink(files)
```

## UTF-16 and UTF-32 support

> Edition two now has much better support for UTF-16 and UTF-32 multi-byte unicode encodings.

Reviewer: Not sure if this is important enough to be worth mentioning in the blog.

> When files with these encodings are read they are automatically converted to UTF-8 internally in an efficeint streaming fashion.

Reviewer 2: ```suggestion
When files with these encodings are read they are automatically converted to UTF-8 internally in an efficient streaming fashion.
```

## Lazy reading

> Edition two uses lazy reading by default.

Reviewer: ```suggestion
Like vroom, the second edition uses lazy reading by default. This means that ...
```

Because readr keeps a file open when reading lazily this means you cannot read, then immediately delete the file.
readr will in most cases close the file once it has been completely read.
However, if you know you want to be able to delete the file after reading it is best to pass `lazy = FALSE` when reading the file.

## Control over quoting and escaping when writing

You can now explicitly control how fields are quoted and escaped when writing with the `quote` and `escape` arguments to `write_*()` functions.

`quote` has three options.

1. 'needed' - which will quote fields only when needed.
2. 'all' - which will always quote all fields.
3. 'none' - which will never quote any fields.

`escape` also has three options, to control how quote characters are escaped.

1. 'double' - which will use double quotes to escape quotes.
2. 'backslash' - which will use a backslash to escape quotes.
3. 'none' - which will not do anything to escape quotes.

We hope these options will give people the flexibility they need when writing files using readr.

## Literal data

In edition one the reading functions treated any input with a newline in it or vectors of length > 1 as literal data.

> In edition two vectors of length > 1 are nowassumed to correspond to multiple files.

Reviewer 2: ```suggestion
In edition two vectors of length > 1 are now assumed to correspond to multiple files.
```

Because of this we now have a more explicit way to represent literal data, by putting `I()` around the input.

```{r}
readr::read_csv(I("a,b\n1,2"))
```

> ## Lighter installation requirements

Reviewer: Maybe lump together some of these smaller changes in a bulleted list under a new heading like "Other minor improvements"

readr now should be much easier to install.
Previous versions of readr used the Boost C++ library to do some of the numeric parsing.
While these are well written, robust libraries, the BH package which contains them has a large number of files (1500+) which can take a long time to install.
In addition the code within these headers is complicated and can take a large amount of memory (2+ Gb) to compile, which made it challenging to compile readr from source in some cases.

readr no longer depends on Boost or the BH package, so should compile more quickly in most cases.

## Deprecated functions and features

* `melt_csv()`, `melt_delim()`, `melt_tsv()` and `melt_fwf()` have been deprecated.
  These functions rely on the first edition parsing code and would be challenging to update to the new parser.

>   When the first edition parsing code is eventually removed from readr they will be split off into a new package.

Reviewer 2: Does this mean the functions are superseded rather than deprecated? My main assumption about deprecated functions is that they will soon no longer work and I need to find an alternative. However we seem to be committed to maintaining a superseded package here.

Reviewer: Yeah, this needs a little tweak of the language.

> * `read_table2()` has been renamed to `read_table()`, as most users expect `read_table()` to work like `utils::read.table()`.

Reviewer: And `read_table2()` has been deprecated?
